packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,119 @@
+import { IResource } from './resource';
+import { Token } from "./token";
+
+/**
+ * Represents the human-readable, physical name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #deployTime()
+ * @see #deployTimeOrAssigned()
+ * @see #of()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically assigned at deploy time.
+   * This is the default if you don't specify a physical name.
+   */
+  public static deployTime(): PhysicalName {
+    return new DeployTimePhysicalName();
+  }
+
+  /**
+   * A physical name that will be assigned at synthesis time if the resource
+   * is used in a cross-environment manner,
+   * and at deploy time otherwise.
+   * This is the default name implicitly created sub-resources
+   * (like IAM Roles) are created with in the Construct Library.
+   * If you're using a Construct in a cross-environment manner,
+   * you need to use either this name,
+   * or a {@link #of() fixed name}.
+   */
+  public static deployTimeOrAssigned(): PhysicalName {
```

Comment:
I actually don't think anymore this is a valid option, and we should only provide `synthTime()`.

The reason being that this option will be unstable between deployments. If I first make a bucket, give it a `deployTimeOrAssigned()` name and deploy, then the name will be CFN generated.

I then extend my CDK app, add a new region, want to read data from the bucket I already have (or whatever) so the bucket gets used in a cross-account manner, and now all of a sudden my bucket gets an auto-calculated name, causing CloudFormation to replace it! 

Or, what's likely to happen in this scenario, CloudFormation creates a new bucket and errors out on deleting the old one (because it's nonempty), leaving you with split state and a stack that requires manual intervention to coax back into a usable state.

In any case, both situations are undesirable.

Reply:
I think this tries to address the "inner resource" concern. You need some way to say "inherit behavior from parent".

Reply:
ping @rix0rrr - can you address my last comment? Thanks!

Reply:
Actually, that's not an issue at all, is it? The original Bucket will simply be orphaned, and the edited Pipeline will use the new one (with the physical name set). So I don't think there's any problem here.